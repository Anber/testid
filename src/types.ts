import * as React from 'react';

export type PropType<
  TElement,
  TExcept extends string
> = TElement extends React.ReactElement<infer TProps>
  ? Omit<TProps, TExcept>
  : {};

export type Id<TAttr extends string> = {
  [P in TAttr]?: string;
};

export type RenderFn<TProps, TAttr extends string> = (
  id: Id<TAttr>
) => React.ReactElement<TProps & Id<TAttr>>;

export interface IAnyProps {
  [prop: string]: any;
}

export interface IConfig<TAttr> {
  display: boolean;
  attr: TAttr;
  separator: string;
}

export type IdComponentProps<
  TChild,
  TAttr extends string
> = TChild extends keyof JSX.IntrinsicElements
  ? JSX.IntrinsicElements[TChild]
  : TChild extends React.ComponentType<infer T>
  ? TAttr extends keyof T
    ? T
    : T extends React.DetailedHTMLProps<any, any>
    ? T
    : never
  : never;

export type IdComponent<TAttr extends string> = <
  TChild extends keyof JSX.IntrinsicElements | React.ComponentType<any>
>(
  props: {
    as: TChild;
    name: string;
  } & IdComponentProps<TChild, TAttr>
) => React.ReactElement;

export type IdHook<TAttr extends string> = (
  name: string,
  relative?: boolean
) => Id<TAttr>;

export type IdsHook<TAttr extends string> = (names: string[]) => Id<TAttr>[];

export type InnerRender<TAttr extends string, TName = string> = <
  TProps extends {}
>(
  name: TName,
  renderFn: RenderFn<TProps, TAttr>
) => React.ReactElement<TProps & Id<TAttr>>;
