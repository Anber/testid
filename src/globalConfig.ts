export interface IConfig {
  display: boolean;
  attr: string;
  separator: string;
}

let globalConfig: IConfig = {
  display: true,
  attr: 'data-test-id',
  separator: ':',
};

export default globalConfig;
