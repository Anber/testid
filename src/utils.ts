export function classNames(
  first?: string | null,
  second?: string | null
): string | undefined {
  if (first && second) {
    return `${first} ${second}`;
  }

  return first || second || undefined;
}
