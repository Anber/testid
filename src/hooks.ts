import * as React from 'react';

import globalConfig from './globalConfig';
import Context from './Context';

export const useRole = (
  name: string,
  relative = name.indexOf(globalConfig.separator) === 0
) => {
  const namespace = React.useContext(Context);
  const role = relative ? `${namespace}${name}` : name;
  return React.useMemo(
    () => ({
      [globalConfig.attr]: role,
    }),
    [globalConfig.attr, role]
  );
};

export const useRoles = (names: string[]) => {
  const namespace = React.useContext(Context);
  return names
    .map(name =>
      name === null || name.indexOf(globalConfig.separator) === 0
        ? `${namespace}${name || ''}`
        : name
    )
    .map(role => ({
      [globalConfig.attr]: role,
    }));
};
