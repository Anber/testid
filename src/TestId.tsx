import * as React from 'react';

import { Consumer, Provider } from './Context';
import globalConfig, { IConfig } from './globalConfig';
import { classNames } from './utils';

type PropType<TElement> = TElement extends React.ReactElement<infer TProps>
  ? Pick<TProps, Exclude<keyof TProps, 'name' | 'children'>>
  : never;

function render<TChild extends React.ReactElement>(
  name: string,
  children: TChild,
  otherProps: PropType<TChild>
) {
  const el = React.cloneElement(
    children,
    Object.assign({}, otherProps, children.props, {
      [globalConfig.attr]: name,
      className:
        classNames(otherProps.className, children.props.className) || undefined,
    })
  );

  return <Provider value={name}>{el}</Provider>;
}

function renderRelative<TChild extends React.ReactElement>(
  name: string,
  children: TChild,
  otherProps: PropType<TChild>
) {
  return (
    <Consumer>
      {namespace => render(`${namespace}${name}`, children, otherProps)}
    </Consumer>
  );
}

function renderWithoutRole<TChild extends React.ReactElement>(
  children: TChild,
  otherProps: PropType<TChild>
) {
  return React.cloneElement(
    children,
    Object.assign({}, otherProps, children.props, {
      className:
        classNames(otherProps.className, children.props.className) || undefined,
    })
  );
}

export default class TestId<
  TChild extends React.ReactElement
> extends React.PureComponent<
  PropType<TChild> & Readonly<{ name: string; children: TChild }>
> {
  static setConfig(config: Partial<IConfig>) {
    Object.assign(globalConfig, config);
  }

  render() {
    const { name, children, ...props } = this.props;
    if (!globalConfig.display) {
      return renderWithoutRole<TChild>(children, props as PropType<TChild>);
    }

    return (name.indexOf(globalConfig.separator) === 0
      ? renderRelative
      : render)(name, children, props as PropType<TChild>);
  }
}
