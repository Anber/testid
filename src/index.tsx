import * as React from 'react';

import Context, { Consumer, Provider } from './Context';
import { classNames } from './utils';
import {
  IAnyProps,
  IConfig,
  IdComponent,
  IdHook,
  IdsHook,
  InnerRender,
  RenderFn,
  Id,
  IdComponentProps,
} from './types';

const mappers = {
  className: classNames,
};

function applyMappers<TOwn extends IAnyProps, TChild extends IAnyProps>(
  ownProps: TOwn,
  childProps: TChild
): IAnyProps {
  const entries = Object.entries(mappers);
  const res: IAnyProps = {};
  for (let [field, mapper] of entries) {
    res[field] = mapper(ownProps[field], childProps[field]);
  }

  return res;
}

function custom<TAttr extends string>(
  config: IConfig<TAttr>
): [IdComponent<TAttr>, IdHook<TAttr>, IdsHook<TAttr>] {
  const render: InnerRender<TAttr> = (name, renderFn) => {
    const idProp: Id<TAttr> = {};
    idProp[config.attr] = name;
    return <Provider value={name}>{renderFn(idProp)}</Provider>;
  };

  const renderRelative: InnerRender<TAttr> = (name, renderFn) => {
    return (
      <Consumer>
        {namespace => render(`${namespace}${name}`, renderFn)}
      </Consumer>
    );
  };

  const renderWithoutId: InnerRender<TAttr, null> = <TProps extends {}>(
    _: null,
    renderFn: RenderFn<TProps, TAttr>
  ) => renderFn({});

  function deprecatedRender({
    children,
    ...otherProps
  }: any): RenderFn<typeof otherProps, TAttr> {
    return id => {
      const resProps: Partial<typeof otherProps & Id<TAttr>> = Object.assign(
        id,
        otherProps,
        children.props,
        applyMappers(otherProps, children.props)
      );

      return React.cloneElement<typeof otherProps & Id<TAttr>>(
        children as any,
        resProps
      );
    };
  }

  function asRender<TProp extends object>(
    as: React.ElementType<TProp>,
    props: TProp,
    ref: React.Ref<HTMLElement>
  ): RenderFn<TProp, TAttr> {
    return id => React.createElement(as, { ...id, ...props, ref });
  }

  const forwardRefCmp = <
    TChild extends keyof JSX.IntrinsicElements | React.ComponentType<any>
  >(
    {
      as = null,
      name,
      ...otherProps
    }: {
      as: TChild;
      name: string;
    } & IdComponentProps<TChild, TAttr>,
    ref: React.Ref<HTMLElement>
  ) => {
    const renderFn: RenderFn<typeof otherProps, TAttr> =
      as !== null
        ? asRender(as, otherProps, ref)
        : deprecatedRender(otherProps);

    if (!config.display) {
      return renderWithoutId<typeof otherProps>(null, renderFn);
    }

    return (name.indexOf(config.separator) === 0 ? renderRelative : render)(
      name,
      renderFn
    );
  };

  const TestId: IdComponent<TAttr> = React.forwardRef(forwardRefCmp) as any; // FIXME

  const useTestId: IdHook<TAttr> = (
    name,
    relative = name.indexOf(config.separator) === 0
  ) => {
    const namespace = React.useContext(Context);
    const id = relative ? `${namespace}${name}` : name;
    return React.useMemo(
      () =>
        ({
          [config.attr]: id,
        } as Id<TAttr>),
      [config.attr, id]
    );
  };

  const useTestIds: IdsHook<TAttr> = names => {
    const namespace = React.useContext(Context);
    return names
      .map(name =>
        name === null || name.indexOf(config.separator) === 0
          ? `${namespace}${name || ''}`
          : name
      )
      .map(
        id =>
          ({
            [config.attr]: id,
          } as Id<TAttr>)
      );
  };

  return [TestId, useTestId, useTestIds];
}

const [TestId, useTestId, useTestIds] = custom({
  attr: 'data-test-id',
  display: true,
  separator: ':',
});

export { IConfig, IdComponent, IdHook, IdsHook, RenderFn, Id };

export { TestId, useTestId, useTestIds, custom };
