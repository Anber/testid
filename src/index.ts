import TestId from './TestId';

export { useRole, useRoles } from './hooks';
export default TestId;
