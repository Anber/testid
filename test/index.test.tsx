import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { custom, Id } from '../src';

const [TestId] = custom({ attr: 'test-id', display: true, separator: ':' });
const Custom: React.FC<{ idx: number; id?: string } & Id<'test-id'>> = ({
  idx,
  ...role
}) => <span {...role}>{idx}</span>;
// @ts-ignore type check
const res = [
  <TestId as={Custom} name="parent" idx={42} />,
  <TestId as="div" name="parent">
    {'something'}
  </TestId>,
];

const run = <TAttr extends string>(
  attr: TAttr,
  separator: string,
  display: boolean
) => {
  const [TestId] = custom({
    attr,
    separator,
    display,
  });

  test(`additional attributes should${
    display ? '' : ' not'
  } be rendered`, () => {
    const component = renderer.create(
      <TestId as={'div'} name="parent">
        <TestId as={'span'} name={`${separator}child`}>
          <TestId as={'i'} name="absolute" />
        </TestId>
      </TestId>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should pass className to the child', () => {
    const component = renderer.create(
      <TestId as={'div'} name="parent" className="test-class" />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should merge classes', () => {
    const component = renderer.create(
      // @ts-ignore deprecated syntax
      <TestId name="parent" className="test-class">
        <div className="existed-class" />
      </TestId>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should forward red', () => {
    function createNodeMock(element: React.ReactElement) {
      if (element.type === 'div') {
        return {};
      }

      return null;
    }

    const ref = React.createRef<HTMLElement>();
    renderer.create(<TestId as="div" name="parent" ref={ref} />, {
      createNodeMock,
    });

    console.log('ref', ref);
    expect(ref.current).not.toBe(null);
  });

  test('should pass unknown props to the child', () => {
    const component = renderer.create(
      // @ts-ignore
      <TestId as={'div'} name="parent" unknownProp="value" />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
};

describe('enabled ids', () => {
  run('data-test-id', ':', true);
});

describe('custom attribute and separator', () => {
  run('data-role', '/', true);
});

describe('disabled ids', () => {
  run('data-test-id', ':', false);
});
