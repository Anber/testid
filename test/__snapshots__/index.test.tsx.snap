// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disabled roles Roles shouldn't be rendered 1`] = `
<div>
  <span>
    <i />
  </span>
</div>
`;

exports[`disabled roles Should merge classes 1`] = `
<div
  className="test-class existed-class"
/>
`;

exports[`disabled roles Should pass className to the child 1`] = `
<div
  className="test-class"
/>
`;

exports[`disabled roles Should pass unknown props to the child 1`] = `
<div
  unknownProp="value"
/>
`;

exports[`enabled roles Roles should be rendered 1`] = `
<div
  data-test-id="parent"
>
  <span
    data-test-id="parent:child"
  >
    <i
      data-test-id="absolute"
    />
  </span>
</div>
`;

exports[`enabled roles Should merge classes 1`] = `
<div
  className="test-class existed-class"
  data-test-id="parent"
/>
`;

exports[`enabled roles Should pass className to the child 1`] = `
<div
  className="test-class"
  data-test-id="parent"
/>
`;

exports[`enabled roles Should pass unknown props to the child 1`] = `
<div
  data-test-id="parent"
  unknownProp="value"
/>
`;
